torch/nn/parallel/distributed.py

```diff
@@ -253,38 +291,75 @@ def _queue_reduction(self, bucket_idx):
             self.reduced = [False] * len(self.bucket_sizes)
 
             def sync_reduction_streams():
-                # We only have to sync with the first one, but it's safer to do it this way
-                # in case we change the way in which we paralellize work
-                r_streams = zip(*self._reduction_streams)
-                for dev_id, default_stream, dev_r_streams in zip(self.device_ids, self._default_streams, r_streams):
-                    with torch.cuda.device(dev_id):
-                        for reduction_stream in dev_r_streams:
-                            default_stream.wait_stream(reduction_stream)
+                # We sync on the default streams for NCCL backend since all
+                # nccl reduction kernels goes to the default streams
+                if dist._backend == "nccl":
+                    for dev_id, default_stream in zip(self.device_ids, self._default_streams):
+                        with torch.cuda.device(dev_id):
+                            default_stream.synchronize()
+                else:
+                    # We only have to sync with the first one, but it's safer to do it this way
+                    # in case we change the way in which we paralellize work
+                    r_streams = zip(*self._reduction_streams)
+                    for dev_id, default_stream, dev_r_streams in zip(self.device_ids, self._default_streams, r_streams):
+                        with torch.cuda.device(dev_id):
+                            for reduction_stream in dev_r_streams:
+                                default_stream.wait_stream(reduction_stream)
+
             Variable._execution_engine.queue_callback(sync_reduction_streams)
 
     def _start_reduction_threads(self):
         num_buckets = len(self.bucket_sizes)
-        self._reduction_queues = [queue.Queue() for _ in range(num_buckets)]
+
+        if dist._backend == "nccl":
+            # Let NCCL backend uses a single Queue for all reduction buckets
+            # in order to maintain the NCCL kernel launch order on different
+            # nodes to avoid the deadlock. In other words, we will only maintain
+            # a single reduction thread for this purpose. This is OK since all
+            # distributed NCCL calls are asynchronous
+            self._reduction_queues = [queue.Queue()]
```

Comment:
@csarofeen I made a single reduction queue and single reduction thread here for NCCL backend.